base_request.py

```diff
@@ -0,0 +1,149 @@
+import requests
+import json
+import resin
+import urllib
+
+from urlparse import urljoin
+
+from .settings import Settings
+from .token  import Token
+from .resources import Message
+
+from exceptions import (
+    JSONDecodeError,
+    NoTokenProvided,
+    ResponseError,
+    RequestError
+)
+
+
+class BaseRequest(object):
+
+    def __init__(self):
+        self.settings = Settings()
+        self.token = Token()
+        self.util = Util()
+
+    def __str__(self):
+        return b'<{:s} at {:#x}>'.format(type(self).__name__, id(self))
+
+    def __unicode__(self):
+        return '<{:s} at {:#x}>'.format(type(self).__name__, id(self))
+
+    def __set_content_type(self, headers, ctype):
+        headers.update({'content-type': ctype})
+
+    def __set_authorization(self, headers):
+        headers.update({'Authorization': 'Bearer {:s}'.format(self.token.get())})
+
+    def __get(self, url, headers, data=None, stream=None):
+        return requests.get(url, headers=headers)
+
+    def __post(self, url, headers, data, stream=None):
+        self.__set_content_type(headers, 'application/json')
+        if not stream:
+            return requests.post(url,data=json.dumps(data), headers=headers)
+        else:
+            return requests.post(url,data=json.dumps(data), headers=headers, stream=stream)
+
+    def __put(self, url, headers, data=None, stream=None):
+        self.__set_content_type(headers, 'application/json')
+        return requests.put(url, data=json.dumps(data), headers=headers)
+
+    def __patch(self, url, headers, data=None, stream=None):
+        self.__set_content_type(headers, 'application/json')
+        return requests.patch(url, data=json.dumps(data), headers=headers)
+
+    def __delete(self, url, headers, data=None, stream=None):
+        self.__set_content_type(headers, 'application/x-www-form-urlencoded')
+        return requests.delete(url, headers=headers)
+
+    def __head(self, url, headers, data=None, stream=None):
+        return requests.head(url, headers=headers)
+
+    def format_params(self, params):
+    	#TODO: make more effecient
+        if params:
+            if params.has_key('expand'):
+                query = "?$expand=" + \
+                        params[
+                            "expand"] + "($filter=" + params['filter'] + "%20eq%20'" + str(params["eq"]) + "')"
+            elif params.has_key('filter'):
+                query = "?$filter=" + params["filter"] + \
+                    "%20eq%20'" + str(params["eq"]) + "'"
```

Comment:
This can be made a lot more readable with [string templates](https://docs.python.org/2/library/string.html#formatstrings).